### DISCUSSION TITLE
Using nanostore with Astro and React

### DISCUSSION THREAD

**Discussion by jolo-dev** [ai/nanostores#105] [Q&A] [answered]
Hey,
Has anyone tried nanostore with [Astro](https://astro.build/)?
My problem is that when using a React component, it seems like `useState` does not seem to work.

Basically, I want to implement a dark mode toggle

```tsx
import React, { useState, useEffect } from "react"
import { Chrono } from "react-chrono";
import { TimelineItemModel } from 'react-chrono/dist/models/TimelineItemModel'
import { Theme } from 'react-chrono/dist/models/Theme'
import { theme } from '../../stores/theme';

function Timeline( { times }) {

    const items: TimelineItemModel[] = times

    const lightTheme: Theme = {
        primary: "#FFF",
        secondary: "#90aab7",
        cardBgColor: "#FFF",
        cardForeColor: "#000",
        titleColor: "#db3853"
    }

    const darkTheme: Theme = {
        primary: "rgb(31,41,55)",
        secondary: "rgb(31,41,55)",
        cardBgColor: "rgb(31,41,55)",
        cardForeColor: "#FFF",
        titleColor: "#db3853"
    }
    const [colors, setColors] = useState<Theme>( lightTheme )

    useEffect( () => {
        theme.get().theme === 'dark' ? setColors( darkTheme ) : setColors( lightTheme )
        console.log( colors );
        theme.listen( ( value, changed ) => {
            if ( value[changed] === 'dark' ) {
                setColors( darkTheme )
                console.log( value[changed] );
                console.log( colors );
            } else {
                setColors( lightTheme )
                console.log( value[changed] );
                console.log( colors );
            }
        } )
    }, [] )

    return (
        <div style={{ width: "85%", height: "85vh", margin: "0 auto" }}>
            <Chrono items={items} lineWidth={10} theme={colors} mode="VERTICAL_ALTERNATING" />
        </div>
    )
}

export default Timeline;
```

The ` console.log( value[changed] );` logs it correctly.
The `theme` from the store also changes successfully but `setColors` does not seem to work.

Anybody has a clue what might be missing?

**Comment by ai** [comment, marked as answer]
In React you should use `useStore` from `@nanostores/react` instead of `useState`/`useEffect`. It will reduce code complexity and may fix the issue.

https://github.com/nanostores/nanostores#react--preact=

**Reply by jolo-dev** [reply]
Hey @ai ,
Thanks for your reply.
Honestly, I don't understand [`nanostores for react`](https://github.com/nanostores/nanostores#react--preact=).
Because it does not listen to changes..

**Reply by ai** [reply]
It should.

Show the code. Maybe there was a bug there?

**Reply by jolo-dev** [reply]
I do

```tsx
const lightTheme: Theme = {
        primary: "#FFF",
        secondary: "#FFF",
        cardBgColor: "#FFF",
        cardForeColor: "#000",
        titleColor: "#db3853"
    }

    const darkTheme: Theme = {
        primary: "rgb(31,41,55)",
        secondary: "rgb(31,41,55)",
        cardBgColor: "rgb(31,41,55)",
        cardForeColor: "#FFF",
        titleColor: "#db3853"
    }
    const store = useStore( theme )
    const colors = store.theme === 'dark' ? darkTheme : lightTheme
```

**Reply by ai** [reply]
Can you show `theme` code?

BTW, `value = useStore(store)` accepts store and return value.

**Reply by jolo-dev** [reply]
```ts
import { persistentMap } from '@nanostores/persistent'

export type SettingsValue = {
  theme: 'dark' | 'light' | 'auto'
}

export const theme = persistentMap('settings:', {
  theme: 'dark'
})
```

**Reply by jolo-dev** [reply]
There is also a toggle in a different component

```ts
theme.setKey('theme', 'dark')
```

**Reply by jolo-dev** [reply]
The `theme.listen()` works. I tested it.

**Reply by ai** [reply]
Hm, everything looks good.

`useStore` uses the same `listen` inside.

Can you show a full file example. How do you define store, the component to listen and how do you change the theme.

**Reply by jolo-dev** [reply]
Let me quickly set up a codesandbox.
Thanks for you help!

**Reply by jolo-dev** [reply]
Hmm... now I cannot make my Toggle to work.

https://codesandbox.io/s/gifted-yonath-6dfwhw?file=/src/components/Toggle.tsx

**Reply by ai** [reply]
1. In `Toggle` you do not need `useState`. Use the store value to get current value.
2. Remove `theme.listen`

**Reply by jolo-dev** [reply]
Well, it still doesn't work :/

**Reply by jolo-dev** [reply]
Now I think it is more an Astro thing rather a nanostore because `useState` does not work properly..

**Reply by ai** [reply]
Yeap, now code looks correctly. You can try this code without Astro to check that it is Astro issue.

**Reply by jolo-dev** [reply]
Yep. It's an Astro thing.
Works on a pure React App.
https://codesandbox.io/s/focused-frost-mu3vxc?file=/src/App.tsx
